### PR TITLE
Add sample for .NET Core IoT with BMP280 Sensor, Azure SDK and Certificates

### PR DESCRIPTION
**Pull request by ikivanc**
This sample application includes implementation of BMP280 sensor with .NET Core IoT on Raspberry Pi, pushes data through AzureSDK using certificates to Azure IoT Hub and save all data to a Azure Blob Storage.

## Description
This is a connected scenario of Raspberry Pi, sensor, Azure with device certificates.

## Motivation and Context
Sample application includes a scenario: Every 3 seconds application reads temperature value from BMP280 sensor and turns on LED after reading the value. Then this temperature value is pushed to IoT Hub with device name and after every successful push LED will turn off. Data will be handled by IoT Hub message routing and will be pushed to blob storage to store.

## How Has This Been Tested?
* This has been tested with .NET Core IoT on Raspberry Pi 2
* Using Adafruit BMP280
* A certificated created manually

## Screenshots
image

image

## Types of changes
- New Sample (adds a new sample)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1665)

### CODE REVIEW

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+﻿using System;
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:1]: You are missing the standard // header. Please look at the other files, you'll find it in any cs file!

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+using System.Device.Gpio;
+using Microsoft.Azure.Devices.Client;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+namespace DNSensorAzureIoTHub
+{
+    public class Program
+    {
+
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:16]: nit: extra line here

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            // set up for LED and pin
+            const int pin = 18;
+            const int lightTime = 1000;
+            const int dimTime = 2000;
+            using GpioController led = new();
+            led.OpenPin(pin, PinMode.Output);
+    
+            // setup for BMP280
+            const int busId = 1;
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:33]: better have all const at the beginning of the function. Not catastrophic for a sample anyway!

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            I2cConnectionSettings i2cSettings = new(busId, Bmp280.DefaultI2cAddress);
+            I2cDevice i2cDevice = I2cDevice.Create(i2cSettings);
+            using var i2CBmp280 = new Bmp280(i2cDevice);
+
+            // Create an X.509 certificate object.
+            var cert = new X509Certificate2($"{DeviceID}.pfx", "1234");
+            var auth = new DeviceAuthenticationWithX509Certificate(DeviceID, cert);
+            DeviceClient azureIoT = DeviceClient.Create(IotBrokerAddress, auth, TransportType.Amqp_Tcp_Only);
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:41]: Why are you using AMQP?

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            if (azureIoT == null)
+            {
+                Console.WriteLine("Failed to create DeviceClient!");
+            }
+            else
+            {
+                Console.WriteLine("Successfully created DeviceClient!");
+            }
+
+            while (true)
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:52]: as you are running in a console, maybe something like `while(!Console.KeyAvailable)` will be better to stop the application. A simple `Console.WriteLine("Press any key to stop the application.");` will be great.

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                    azureIoT.SendEventAsync(eventMessage).Wait();
+                    Console.WriteLine($"Data is pushed to Iot Hub: {message}");
+
+                    // led off
+                    led.Write(pin, PinValue.Low);
+                    Thread.Sleep(dimTime);
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"An error occured: {ex.Message}");
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:84]: As a bonus, you may blink the led fast 3 times? :-)

File: samples/bmp280-sensor-azure-iot-hub/Readme.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
+# Raspberry Pi with .NET Core IoT
+
+Raspberry Pi is one of the most popular boards and now .NET Core IoT can be used on top of Raspberry Pi OS to enable the writing of managed code applications.
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Readme.md:3]:
```suggestion
Raspberry Pi is one of the most popular boards and now .NET IoT can be used on top of Raspberry Pi OS to enable the writing of managed code applications.
```

File: samples/bmp280-sensor-azure-iot-hub/Readme.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+## Remote Debugging
+
+To enable remote debugging on Raspberry Pi make sure following below steps.
+
+* [Remote debugging with VS Code on Windows to a Raspberry Pi using .NET Core on ARM](https://www.hanselman.com/blog/remote-debugging-with-vs-code-on-windows-to-a-raspberry-pi-using-net-core-on-arm)
+
+* [How to use Windows 10's built-in OpenSSH to automatically SSH into a remote Linux machine](https://www.hanselman.com/blog/how-to-use-windows-10s-builtin-openssh-to-automatically-ssh-into-a-remote-linux-machine)
+
+After the setup we have [launch.json](../.vscode/launch.json) file to enable build and publish steps. Also we have [tasks.json](../.vscode/tasks.json) . file to enable remote debugging feature helps us to debug our application on Raspberry Pi on Raspberry Pi OS with .NET Core IoT.
```
Comment (Ellerbach) [inline samples/bmp280-sensor-azure-iot-hub/Readme.md:94]: Would be great to add as well for Visual Studio. A lot of developers are directly using Visual Studio.

**Review by Ellerbach — COMMENTED** [review]
Thanks a lot for this sample, it's great to have it detailed, step by step!
All up, the sample looks good to me!
Please fix the build issues and let's have few more people reviewing this content.

**Comment by CamSoper** [comment]
This looks great! Very similar to a MS Learn module I'm writing.

One nit: Instead of `dotnet/iot`, can you please put this in `dotnet/samples`? That way if we put the proper metadata in the README, it'll automatically get published in the .NET Samples Browser. 

I can help with that, though probably not until after next week.

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 1 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:1]: Sure, added, thanks

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 2 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:16]: Fixed

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 3 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:33]: Moved to top, thanks

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 5 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:52]: Great point, change with this

File: eng/Version.Details.xml [hunk 9]
```diff
@@ -1,17 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Dependencies>
   <ToolsetDependencies>
-    <Dependency Name="Microsoft.DotNet.Arcade.Sdk" Version="7.0.0-beta.21463.4">
+    <Dependency Name="Microsoft.DotNet.Arcade.Sdk" Version="7.0.0-beta.21466.4">
```
Comment (krwq) [inline eng/Version.Details.xml:4]: I don't believe these should be part of this PR

File: samples/bmp280-sensor-azure-iot-hub/publish.bat [hunk 10]
```diff
@@ -0,0 +1,4 @@
+dotnet publish -r linux-arm /p:ShowLinkerSizeComparison=true
```
Comment (krwq) [inline samples/bmp280-sensor-azure-iot-hub/publish.bat:1]: I'd recommend to add this file somewhere in he readme instead

File: src/devices/Mcp960x/Mcp960x.csproj [hunk 11]
```diff
@@ -12,7 +12,6 @@
   <ItemGroup>
     <Compile Include="*.cs" />
     <None Include="README.md" />
```
Comment (krwq) [inline src/devices/Mcp960x/Mcp960x.csproj:14]: I think this shouldn't be part of this PR - presumably you did some kind of pull + rebase on your branch if this wasn't intentional

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 6 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:91]: Nice suggestion, this is added

File: samples/bmp280-sensor-azure-iot-hub/Program.cs (hunk 4 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:47]: There's no reason to pick this specifically, I changed to MQTT protocol

File: samples/bmp280-sensor-azure-iot-hub/Readme.md (hunk 8 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/Readme.md:94]: This is added as well.

File: samples/bmp280-sensor-azure-iot-hub/publish.bat (hunk 10 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/publish.bat:1]: Thanks for your comment @krwq, this .bat file is used by vscode tasks.json. What do you think, should I add this just in readme file?

File: src/devices/Mcp960x/Mcp960x.csproj (hunk 11 above)
Comment (ikivanc) [reply src/devices/Mcp960x/Mcp960x.csproj]: Fixed this issue

File: eng/Version.Details.xml (hunk 9 above)
Comment (ikivanc) [reply eng/Version.Details.xml]: This is fixed

File: samples/bmp280-sensor-azure-iot-hub/publish.bat (hunk 10 above)
Comment (ikivanc) [reply samples/bmp280-sensor-azure-iot-hub/publish.bat]: removed `publish.bat` and added into readme file as instruction.

**Comment by Ellerbach** [comment]
@rido-min for you as well :-)

File: global.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "6.0.100-rc.1.21430.12",
+    "dotnet": "6.0.100-rc.1.21458.32",
```
Comment (krwq) [inline global.json:3]: this will need to be reverted

File: src/devices/Mcp960x/Mcp960x.csproj (hunk 11 above)
Comment (krwq) [reply src/devices/Mcp960x/Mcp960x.csproj:14]: seems still shows up on the PR, also below

File: global.json (hunk 12 above)
Comment (krwq) [reply global.json:3]: (seperate PR for unrelated changes)

Comment (krwq) [reply global.json:3]: @joperezr this seems to be making the CI green again

**Comment by ikivanc** [comment]
Hi @joperezr, is it possible to take a look at the engineering changes in this PR?

**Comment by joperezr** [comment]
> is it possible to take a look at the engineering changes in this PR?

Those changes we usually don't make automatically, instead they get created and pushed automatically by Arcade. Can you please ellaborate why you need new Arcade/Helix sdks as well as new dotnet sdks?

File: global.json (hunk 12 above)
Comment (ikivanc) [reply global.json]: I revert back and it is successfully passed

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+namespace DNSensorAzureIoTHub
+{
+    public class Program
+    {
+        // set up IoT Hub message
+        const string DeviceID = "<replace-with-your-device-id>";
+        const string IotBrokerAddress = "<replace-with-your-iot-hub-name>.azure-devices.net";
+
+        // LED constraints 
+        const int pin = 18;
```
Comment (krwq) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:24]: for consts please start with capital letter
```suggestion
        private const int Pin = 18;
```

(+ changes to all references)

also you should specify visibility explicitly (as above)

File: samples/bmp280-sensor-azure-iot-hub/Program.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        // busId number for I2C pins
+        const int busId = 1;
+
+        public static void Main()
+        {
+            Console.WriteLine($".Net IoT with BMP280 Sensor!");
+
+            // set up for LED and pin
+            using GpioController led = new();
+            led.OpenPin(pin, PinMode.Output);
```
Comment (krwq) [inline samples/bmp280-sensor-azure-iot-hub/Program.cs:37]: @Ellerbach @raffaeler @pgrawehr @joperezr interesting usage of GpioController. Feels like `Pin` class could be useful

Comment (krwq) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:37]: (I mean the naming of GpioController suggests it's treated as an led rather than for managing pins)

Comment (raffaeler) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:37]: Agree, we should definitely discuss this.
My first thought is to share the private `_openPins` dictionary (making it static?) so that you can control the pin either seprately or with the controller.
Probably the events logic should be revised as well.

In general, I find the ability to use single pins interesting and more intuitive for new users.

Comment (joperezr) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:37]: We can have this discussion in an issue instead. The original design of the API was very much considering that we wouldn't have a Pin class, so now that we have gone through that route, a lot would have to change if we were to introduce it now.

Comment (raffaeler) [reply samples/bmp280-sensor-azure-iot-hub/Program.cs:37]: I just created a discussion [here](https://github.com/dotnet/iot/issues/1671) to track this.

File: src/devices/Mcp960x/Mcp960x.csproj (hunk 11 above)
Comment (joperezr) [reply src/devices/Mcp960x/Mcp960x.csproj:14]: It is fine to keep, in fact I believe not doing so would break the build so I don't mind if this fix is also checked in.

**Comment by rido-min** [comment]
Have you considered to make this sample PnP compatible? It will require: 1) define the model with the telemetry being sent and 2) announce the model id in the connection.

**Review by Ellerbach — APPROVED** [review]
Please resolve the conflicts and I guess we will be good to go!
